File: ts/project/.sdk/tm/lua/utility/make_options.lua
-- ProjectName SDK utility: make_options

local vs = require("utility.struct.struct")
local schema = require("schema")

-- See the call site in make_options for why this exists.
local function densify_lists(v, seen)
  if type(v) ~= 'table' then
    return v
  end

  seen = seen or {}
  if seen[v] then
    return v
  end
  seen[v] = true

  local count, max, intonly = 0, 0, true
  for k in pairs(v) do
    count = count + 1
    if math.type(k) == 'integer' and 0 < k then
      if k > max then
        max = k
      end
    else
      intonly = false
    end
  end

  -- A hole is the only case worth rebuilding for: every key an integer, and
  -- a highest key past the number of entries.
  if intonly and 0 < count and max > count then
    local out = {}
    for i = 1, max do
      local e = v[i]
      if e == nil then
        out[i] = false
      else
        out[i] = densify_lists(e, seen)
      end
    end
    return out
  end

  for k, e in pairs(v) do
    v[k] = densify_lists(e, seen)
  end
  return v
end

local function copy_data(v)
  if type(v) ~= 'table' then
    return v
  end
  local out = {}
  for k, e in pairs(v) do
    out[k] = copy_data(e)
  end
  return setmetatable(out, getmetatable(v))
end

-- struct's lua getprop scans every key of the parent to classify it, so
-- validating a map of n entities costs n^2. Each entity entry validates on its
-- own, and the config's entries are fixed for this Lua state: validate them
-- once, and per client only the entries the caller passes.
local entity_defaults_src = nil
local entity_defaults = nil

local function validate_entities(entities, spec)
  local out = vs.validate({ entity = densify_lists(entities) }, { entity = spec })
  if type(out) == 'table' and type(out.entity) == 'table' then
    return out.entity
  end
  return {}
end

local function validated_entity_defaults(cfgentity, spec)
  if entity_defaults_src ~= cfgentity then
    entity_defaults = validate_entities(vs.clone(cfgentity), spec)
    entity_defaults_src = cfgentity
  end
  return entity_defaults
end

local function make_options_util(ctx)
  local options = ctx.options or {}

  -- Merge utility overrides from options onto the utility object.
  --
  -- A key naming a real utility member REPLACES it; anything else is attached
  -- as a custom extra. Shelving everything in `custom` - a table nothing reads
  -- - made `utility = { fetcher = ... }`, the documented transport seam, a
  -- silent no-op here while ts honoured it.
  --
  -- Option keys are camelCase, as ts spells them; members here are
  -- snake_case. Only a PUBLIC name may replace: public utility names carry no
  -- underscore, so an underscore means the caller named something of their
  -- own - possibly the internal spelling of a real member. `make_error` must
  -- stay an extension, or a non-callable would break the error path on the
  -- next request.
  local custom_utils = vs.getprop(options, "utility")
  if type(custom_utils) == "table" then
    local utility = ctx.utility
    if utility ~= nil then
      for key, val in pairs(custom_utils) do
        local member = nil
        if type(key) == "string" and not key:find("_") then
          member = key:gsub("(%u)", function(c) return "_" .. c:lower() end)
        end
        if member ~= nil and member ~= "custom" and utility[member] ~= nil then
          utility[member] = val
        else
          utility.custom[key] = val
        end
      end
    end
  end

  local opts = vs.clone(options)
  if type(opts) ~= "table" then
    opts = {}
  end

  -- Feature add-order. options.feature may be given as an ordered LIST of
  -- { name = ..., active = ..., ... } entries (the list position IS the order
  -- in which features are added), or as a { name = {opts} } map. Normalize a
  -- list to a map (so merge/validate/init are unchanged) and remember the
  -- explicit order; a map defaults to test-first so the `test` mock transport
  -- is installed as the base of the transport wrapper chain.
  local featureorder = {}
  if vs.islist(opts.feature) then
    local fmap = {}
    for _, entry in ipairs(opts.feature) do
      if type(entry) == "table" and entry.name ~= nil then
        local name = entry.name
        local fopts = {}
        for k, v in pairs(entry) do
          if k ~= "name" then
            fopts[k] = v
          end
        end
        fmap[name] = fopts
        table.insert(featureorder, name)
      end
    end
    opts.feature = fmap
  end

  local config = ctx.config or {}
  local cfgopts = {}
  local co = config["options"]
  if type(co) == "table" then
    cfgopts = co
  end

  -- THE OPTION SPEC IS GENERATED, NOT WRITTEN HERE.
  --
  -- `schema.OPTSPEC` is built from the model: `main.kit.optspec` for
  -- the standard options, plus one entry per feature this target
  -- carries, taken from that feature's own `config.options` /
  -- `config.optspec`. Editing this file to add an option would put it
  -- back where it was — one of twenty hand-maintained copies of a
  -- schema nothing cross-checked — so add it to the model instead and
  -- every ported target validates it.
  --
  -- NOT MUTATED. It is a module-level table shared by every client this
  -- Lua state constructs; anything defaulted below is applied to the
  -- RESULT, never to the spec.
  local optspec = schema.OPTSPEC

  -- Preserve system.fetch before merge/validate.
  local sys_fetch = vs.getpath(opts, "system.fetch")

  -- Preserve extend feature INSTANCES before merge/validate, by ORIGINAL
  -- reference (the ts makeOptions keeps the caller's instances too):
  -- merge/validate are data-oriented and would mangle instance tables
  -- carrying functions -- the same reason system.fetch is preserved.
  local extend = nil
  if type(options) == "table" and type(options["extend"]) == "table" then
    extend = options["extend"]
  end

  -- Clone the config side before merging: `config` is a per-Lua-state
  -- singleton (see config_shared), and merge would otherwise use its nested
  -- tables as merge TARGETS — one instance's options (server, headers, ...)
  -- would contaminate every instance constructed after it.
  local cfgentity = cfgopts["entity"]
  local userentity = opts["entity"]
  local split = type(cfgentity) == 'table' and vs.ismap(cfgentity)
    and (userentity == nil or vs.ismap(userentity))

  if split then
    local rest = {}
    for k, v in pairs(cfgopts) do
      if k ~= "entity" then
        rest[k] = v
      end
    end
    cfgopts = rest
    opts["entity"] = nil
  end

  local merged = vs.merge({ {}, vs.clone(cfgopts), opts })

  -- LUA CANNOT STORE NIL, so `{ a, nil, b }` is a table holding keys 1 and 3
  -- and no 2 -- not a sequence. struct classifies it as a MAP and refuses it
  -- against a `list` spec, so a caller's sparse list became an uncatchable
  -- construction error naming the whole option shape, instead of the feature's
  -- own message about the one entry that is wrong.
  --
  -- Densify first: a table whose keys are ALL positive integers is meant as a
  -- list, so walk 1..max and fill each hole with `false`. The hole stays
  -- PRESENT -- dropping it would silently shorten a provider chain, which is
  -- the failure the secrets suite pins -- and `false` is not a table, so the
  -- feature reading it refuses it by its own rule, with its own message.
  merged = densify_lists(merged)

  local validated = vs.validate(merged, optspec)
  if type(validated) ~= "table" then
    validated = {}
  end
  opts = validated

  if split then
    local entity = copy_data(validated_entity_defaults(cfgentity, optspec["entity"]))
    if userentity ~= nil then
      local picked = {}
      for name, _ in pairs(userentity) do
        picked[name] = cfgentity[name]
      end
      local own = vs.merge({ {}, vs.clone(picked), userentity })
      for name, e in pairs(validate_entities(own, optspec["entity"])) do
        entity[name] = e
      end
    end
    opts["entity"] = entity
  end

  -- Resolve a templated base URL (e.g. https://{tenant_id}.hanko.io).
  -- Every placeholder must resolve to a non-empty value: from
  -- options.server (user), else the config default. A placeholder that
  -- resolves to "" is a construction ERROR in live mode — the URL cannot
  -- work — but in test mode substitutes the deterministic value
  -- "test-<name>" so offline tests need no configuration.
  local base = opts.base
  if type(base) == "string" and string.find(base, "{", 1, true) then
    local testmode = vs.getpath(opts, "test.active") == true
      or vs.getpath(opts, "feature.test.active") == true
    local server = type(opts.server) == "table" and opts.server or {}
    local sdkname = vs.getpath(config, "main.name")
    if type(sdkname) ~= "string" or sdkname == "" then
      sdkname = "SDK"
    end
    opts.base = string.gsub(base, "{([%w_]+)}", function(name)
      local val = server[name]
      if type(val) ~= "string" then
        val = ""
      end
      if val == "" then
        if testmode then
          return "test-" .. name
        end
        error(sdkname .. ": the server variable '" .. name .. "' is required: " ..
          "the API base URL is '" .. base .. "' — pass " ..
          "{ server = { " .. name .. " = \"...\" } } in the SDK options", 0)
      end
      return val
    end)
  end

  -- Restore system.fetch.
  if sys_fetch ~= nil then
    if type(opts["system"]) == "table" then
      opts["system"]["fetch"] = sys_fetch
    else
      opts["system"] = { fetch = sys_fetch }
    end
  end

  -- Restore extend feature instances.
  if extend ~= nil then
    opts["extend"] = extend
  end

  -- Derived clean config.
  local clean_keys = "key,token,id"
  local ck = vs.getpath(opts, "clean.keys")
  if type(ck) == "string" then
    clean_keys = ck
  end

  local parts = {}
  for part in string.gmatch(clean_keys, "[^,]+") do
    local trimmed = part:match("^%s*(.-)%s*$")
    if trimmed ~= "" then
      table.insert(parts, vs.escre(trimmed))
    end
  end
  local keyre = table.concat(parts, "|")

  -- Resolve the feature add-order: an explicit list order (above) wins;
  -- otherwise order the map test-first, then the remaining names sorted, so
  -- the outcome is deterministic and `test` is always the base transport.
  if #featureorder == 0 then
    local fmap = opts.feature
    local names = {}
    if type(fmap) == "table" then
      for k, _ in pairs(fmap) do
        if type(k) == "string" then
          table.insert(names, k)
        end
      end
    end
    table.sort(names)
    local has_test = false
    for _, n in ipairs(names) do
      if n == "test" then
        has_test = true
      end
    end
    if has_test then
      featureorder = { "test" }
      for _, n in ipairs(names) do
        if n ~= "test" then
          table.insert(featureorder, n)
        end
      end
    else
      featureorder = names
    end
    -- Station special case, mirroring test's: its transport wrap must
    -- sit immediately outside the base transport (inside retry/cache/
    -- netsim), so map-form activation hoists it to just after test -
    -- or first, when no test entry exists. Without this the sorted
    -- default would init station last and wrap OUTSIDE the recording
    -- features, turning its wire-truth events into fiction.
    local si = nil
    for i, n in ipairs(featureorder) do
      if n == "station" then
        si = i
        break
      end
    end
    if si ~= nil then
      table.remove(featureorder, si)
      local ti = 0
      for i, n in ipairs(featureorder) do
        if n == "test" then
          ti = i
          break
        end
      end
      table.insert(featureorder, ti + 1, "station")
    end
  end

  local derived = { clean = {} }
  if keyre ~= "" then
    derived.clean = { keyre = keyre }
  end
  derived.featureorder = featureorder
  opts["__derived__"] = derived

  return opts
end

return make_options_util

File: ts/project/.sdk/tm/go/utility/make_options.go
package utility

import (
	"encoding/json"
	"regexp"
	"sort"
	"strings"

	vs "github.com/voxgig/struct"

	"GOMODULE/core"
)

// {name} placeholders in a templated server URL (OpenAPI server variables).
var serverVarRe = regexp.MustCompile(`\{[A-Za-z0-9_]+\}`)

func canonNumbers(node any) {
	switch n := node.(type) {
	case map[string]any:
		for k, v := range n {
			if num, ok := v.(json.Number); ok {
				n[k] = canonNumber(num)
			} else {
				canonNumbers(v)
			}
		}
	case []any:
		for i, v := range n {
			if num, ok := v.(json.Number); ok {
				n[i] = canonNumber(num)
			} else {
				canonNumbers(v)
			}
		}
	}
}

func canonNumber(n json.Number) any {
	if i, err := n.Int64(); err == nil {
		return i
	}
	if f, err := n.Float64(); err == nil {
		return f
	}
	return n
}

func makeOptionsUtil(ctx *core.Context) map[string]any {
	options := ctx.Options
	if options == nil {
		options = map[string]any{}
	}

	if customUtils := core.ToMapAny(options["utility"]); customUtils != nil {
		utility := ctx.Utility
		if utility != nil {
			for key, val := range customUtils {
				if !overrideUtil(utility, key, val) {
					utility.Custom[key] = val
				}
			}
		}
	}

	authval, authgiven := options["auth"]
	authsuppressed := authgiven && authval == nil

	opts := vs.Clone(options).(map[string]any)

	var featureorder []any
	if farr, ok := opts["feature"].([]any); ok {
		fmap := map[string]any{}
		for _, entry := range farr {
			em := core.ToMapAny(entry)
			if em == nil {
				continue
			}
			name, _ := em["name"].(string)
			if name == "" {
				continue
			}
			fopts := map[string]any{}
			for k, v := range em {
				if k != "name" {
					fopts[k] = v
				}
			}
			fmap[name] = fopts
			featureorder = append(featureorder, name)
		}
		opts["feature"] = fmap
	}

	config := ctx.Config
	if config == nil {
		config = map[string]any{}
	}
	cfgopts := map[string]any{}
	if co, ok := config["options"]; ok && co != nil {
		if cm, ok := co.(map[string]any); ok {
			cfgopts = cm
		}
	}

	optspec := core.OPTSPEC

	// Preserve system.fetch before merge/validate.
	var sysFetch any
	if sf := vs.GetPath(opts, []any{"system", "fetch"}); sf != nil {
		sysFetch = sf
	}

	// Clone the config side before merging: `config` is a process-wide
	// singleton (see core.SharedConfig), and Merge would otherwise use its
	// nested maps as merge TARGETS — one instance's options (server, headers,
	// ...) would contaminate every instance constructed after it.
	merged := vs.Merge([]any{map[string]any{}, vs.Clone(cfgopts), opts})

	// To reflect a json.Number is a string, so the validator refuses one where
	// the feature readers accept it. Canonicalise first: options decoded with
	// UseNumber() are otherwise judged by a rule nothing else applies.
	canonNumbers(merged)

	validated, _ := vs.Validate(merged, optspec)
	opts = validated.(map[string]any)

	// Restore the suppression the optspec default would otherwise erase.
	if authsuppressed {
		opts["auth"] = nil
	}

	if base, ok := opts["base"].(string); ok && strings.Contains(base, "{") {
		testmode := false
		if ta, ok := vs.GetPath(opts, []any{"test", "active"}).(bool); ok && ta {
			testmode = true
		}
		if fa, ok := vs.GetPath(opts, []any{"feature", "test", "active"}).(bool); ok && fa {
			testmode = true
		}
		server := core.ToMapAny(opts["server"])
		sdkname := "SDK"
		if mn, ok := vs.GetPath(config, []any{"main", "name"}).(string); ok && mn != "" {
			sdkname = mn
		}
		resolved := serverVarRe.ReplaceAllStringFunc(base, func(ph string) string {
			name := ph[1 : len(ph)-1]
			val, _ := server[name].(string)
			if val == "" {
				if testmode {
					return "test-" + name
				}
				panic(sdkname + ": the server variable '" + name + "' is required: " +
					"the API base URL is '" + base + "' — pass " +
					`options["server"].(map)["` + name + `"] in the SDK options`)
			}
			return val
		})
		opts["base"] = resolved
	}

	// Restore system.fetch.
	if sysFetch != nil {
		if sys, ok := opts["system"]; ok {
			if sm, ok := sys.(map[string]any); ok {
				sm["fetch"] = sysFetch
			}
		} else {
			opts["system"] = map[string]any{"fetch": sysFetch}
		}
	}

	// Derived clean config.
	cleanKeys := "key,token,id"
	if ck := vs.GetPath(opts, []any{"clean", "keys"}); ck != nil {
		if cks, ok := ck.(string); ok {
			cleanKeys = cks
		}
	}

	parts := strings.Split(cleanKeys, ",")
	var filtered []string
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p != "" {
			filtered = append(filtered, vs.EscRe(p))
		}
	}
	keyre := strings.Join(filtered, "|")

	// Resolve the feature add-order: an explicit array order (above) wins;
	// otherwise order the map test-first, then the remaining names sorted, so
	// the outcome is deterministic and `test` is always the base transport.
	if len(featureorder) == 0 {
		fmap := core.ToMapAny(opts["feature"])
		names := make([]string, 0, len(fmap))
		for k := range fmap {
			names = append(names, k)
		}
		sort.Strings(names)
		hasTest := false
		for _, n := range names {
			if n == "test" {
				hasTest = true
			}
		}
		ordered := make([]string, 0, len(names))
		if hasTest {
			ordered = append(ordered, "test")
			for _, n := range names {
				if n != "test" {
					ordered = append(ordered, n)
				}
			}
		} else {
			ordered = append(ordered, names...)
		}
		si := -1
		for i, n := range ordered {
			if n == "station" {
				si = i
				break
			}
		}
		if si >= 0 {
			ordered = append(ordered[:si], ordered[si+1:]...)
			ti := 0
			for i, n := range ordered {
				if n == "test" {
					ti = i + 1
					break
				}
			}
			ordered = append(ordered[:ti],
				append([]string{"station"}, ordered[ti:]...)...)
		}
		for _, n := range ordered {
			featureorder = append(featureorder, n)
		}
	}

	derived := map[string]any{
		"clean": map[string]any{},
	}
	if keyre != "" {
		derived["clean"] = map[string]any{"keyre": keyre}
	}
	derived["featureorder"] = featureorder
	opts["__derived__"] = derived

	return opts
}
